numpy/core/numeric.py

```diff
@@ -623,6 +624,66 @@ def asfortranarray(a, dtype=None):
     """
     return array(a, dtype, copy=False, order='F', ndmin=1)
 
+def pack_fields(a):
+    """Return copy of a structured array with fields "packed" without padding.
+
+    The memory layout of structured datatypes allows fields at arbitrary
+    byte offsets. This means the fields can be separated by padding bytes,
+    their offsets can be non-monotonically increasing, and they can overlap.
+
+    This method repacks the fields to produce a memory layout in which
+    each field starts at the byte the previous field ended, with increasing
+    byte offsets.
+
+    This method also accepts a dtype object, in which case it returns
+    a dtype with repacked fields.
```

Comment:
I wonder if we want to name this `dtype.packed_fields()`, or some method of dtype objects

Reply:
Yeah that would avoid polluting the `np` namespace.  

But I do have an aesthetic objection: I don't like that OO style, of making it the attribute of a class: Being used to C, I think toplevel functions are more pleasing than (class) methods, and my impression is numpy is usually organized that way too. Most other dtype related functions are functions, not dtype methods: https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.dtype.html. I always use `np.sum(a)`, not `a.sum()`. 

`pack_fields` also happens to be a "pure function" in the functional programming sense, somehow justifying to me it should not be a class method. I can support putting it in a namespace (not a class) like `np.lib.recfunctions`, `np.rec` or somewhere else though.

PS. I am going to rename it `repack_fields(a, align=False)`.

Reply:
> Most other dtype related functions